ls1intum/Artemis
# `General`: Improve the retrieval of user data when authenticating over external service providers based on SAML2

### Checklist
#### General


     (Not possible, as SAML2 authentication is not configured on the test servers.)
- [x] This is a small issue that I tested locally and works on our production server ~~and was confirmed by another developer on a test server~~.
#### Server

### Motivation and Context
Our SAML2 provider changed their configuration in a way such that the registration number now additionally includes some prefix instead of only the actual number. The admin cannot change this on their end. Therefore, the value of interest has to be extracted from the whole attribute value when received from Artemis.

### Description
This PR introduces a new optional configuration parameter `saml2.value-extraction-patterns`. It contains a list of key-value pairs of the format:
- `key`: The name of the SAML2 attribute for which the received value needs some additional extraction.
- `value-pattern`: This is a regular expression that has to match the whole attribute value. It has to contain a named capture group. Only the part of the received value that is matched by this group is stored in Artemis.

If either the key pattern or the value pattern do not match the received attribute, then no transformation is applied to the value. Instead, the value is used as received from the authentication provider.

### Steps for Testing
Only testing with the local installation is possible as the test servers do not have SAML2 logins configured.

1. Artemis should start without the `saml2` profile.

For the following steps, start your local Artemis with the `saml2` profile enabled. You may have to uncomment the example `metadata`, `entity-id`, and `registration-id` in the `identity-providers` list for that to work.
1. Do not enable any of the extraction key/value pairs in `src/main/resources/config/application-saml2.yml`. Artemis should start normally.
2. Enable just one of the two elements of a pair. Artemis should now fail to start.
3. Enable both key and value_pattern. Artemis should now start again.
4. Modify the value of a `value_pattern` such that it no longer contains a named capture group with the name `value`. Artemis should fail to start.
5. Fix the capture group name. Enable an additional key/value_pattern pair with the same key. Artemis should fail to start.

The actual functionality is hard to test without a proper SAML2 provider set up. If you have:
1. Not defining any key/value pairs should not change the behaviour when authenticating via SAML2.
2. Choose a key/value pair where you can extract some part of the value using a regex. Check that only the part of the value that is matched by the capture group is used by Artemis.

### Review Progress
#### Code Review
#### Manual Tests

### Test Coverage
| Class | Branch | Line |
| --- | --- | --- |
| `SAML2Service` | 75% | 93% |
